README.md
usrp-utils
==========

Some UHD level utility programs for use with USRP devices:


This software includes the following utilities:

	usrp-energy
	usrp-record
	usrp-sensor
        energycalculator
	fftcompute



Documentation for usrp-record:
 *This program is designed to record the samples streamed from
 *a USRP device.  This records the raw 16-bit I/Q samples from the USRP
 *
 *The commandline options are:
 *
 * -o [file]       Output File  -The output file contains raw float data
 *                               representing the computed spectral periodigram
 *                               of the recorded signal.  The peridigram is
 *                               computed with fftw3f's 1D DFT.
 *
 * -a [args]       USRP Args    -Specify the address for the input USRP. See
 *               http://files.ettus.com/uhd_docs/manual/html/identification.html
 *                               for information about identifying USRPs.
 *
 * -f [frequency]  Center Freq  -The center frequency for the FFT process.
 *                               Future plans including adding sweep
 *                               functionality to the program.
 *
 * -r [rate]       Sample Rate  -The sample rate of the USRP.  Because of the
 *                               I-Q data from the USRP, this correlates to the
 *                               bandwidth of the FFT calculation.
 *
 * -t [time]       Runtime      -The runtime in seconds for the sensing process.
 *
 *
 * -g [gain]       RX Gain      -Gain in DB of the rx chain



Documentation for usrp-energy:

 *This program is designed to record the magnitude of samples streamed from
 *a USRP device.
 *
 *The commandline options are:
 *
 * -o [file]       Output File  -The output file contains raw float data
 *                               representing the computed spectral periodigram
 *                               of the recorded signal.  The peridigram is
 *                               computed with fftw3f's 1D DFT.
 *
 * -b [size]       Bin Size     -Energy bin size in samples
 *
 * -a [args]       USRP Args    -Specify the address for the input USRP. See
 *               http://files.ettus.com/uhd_docs/manual/html/identification.html
 *                               for information about identifying USRPs.
 *
 * -f [frequency]  Center Freq  -The center frequency for the FFT process.
 *                               Future plans including adding sweep
 *                               functionality to the program.
 *
 * -r [rate]       Sample Rate  -The sample rate of the USRP.  Because of the
 *                               I-Q data from the USRP, this correlates to the
 *                               bandwidth of the FFT calculation.
 *
 * -t [time]       Runtime      -The runtime in seconds for the sensing process.
 *
 *
 * -g [gain]       RX Gain      -Gain in DB of the rx chain



Documentation for usrp-sensor:

 *This program is designed to leverage single-precision fftw3 libraries and
 *pthreads to compute spectral periodigrams using an applied window function
 *and the magnitude of a 1D DFT.  A USRP serves as the input data stream for
 *this program.  As of July 2012, this has only been built and tested with an
 *N210 and SBX RX/TX daughtercard.
 *
 *The commandline options are:
 *
 * -o [file]       Output File  -The output file contains raw float data
 *                               representing the computed spectral periodigram
 *                               of the recorded signal.  The peridigram is
 *                               computed with fftw3f's 1D DFT.
 *
 * -s [size]       FFT Size     -Size of the FFT to compute
 *
 * -l [overlap]    Overlap      -How many FFTs overlap in the space of a
 *                               transform.  That is, a transform of FFT Size is
 *                               computed every (FFT Size)/Overlap samples.
 *                               Sanity checking is performed to ensure that the
 *                               speficied overlap is appropriate
 *
 * -w [file]       Window File  -The window file is expected to contain raw
 *                               floats representing the window function to
 *                               apply to each FFT.  It is acceptable for the
 *                               window function to be smaller than the FFT.
 *                               If this is the case, then the program
 *                               left-justifies the window with respect to the
 *                               input array to fftw3f, and zero-pads the
 *                               remaining values.
 *
 * -c [children]   Child Threads-This specifies the number of child worker
 *                               threads to spawn for .  If your
 *                               processor has N cores (including hyperthreading
 *                               cores), it is recommended to specify N-1
 *                               children.  This allows for the parent thread
 *                               to process the input data for a total of N
 *                               threads.
 *
 * -a [args]       USRP Args    -Specify the address for the input USRP. See
 *               http://files.ettus.com/uhd_docs/manual/html/identification.html
 *                               for information about identifying USRPs.
 *
 * -f [frequency]  Center Freq  -The center frequency for the FFT process.
 *                               Future plans including adding sweep
 *                               functionality to the program.
 *
 * -r [rate]       Sample Rate  -The sample rate of the USRP.  Because of the
*                               I-Q data from the USRP, this correlates to the
 *                               bandwidth of the FFT calculation.
 *
 * -t [time]       Runtime      -The runtime in seconds for the sensing process.
 *
 * -g [gain]       RX Gain      -Gain in DB of the rx chain
 *
 *Description of error messages:
 *
 *Need at least one child thread
 *  Worker threads spawn from the parent process.  You must specify at least
 *  one child thread to do the FFT calculations.
 *
 *Cannot open window file
 *  There was a problem opening the window file.
 *
 *Incompatible FFT Size and Overlap factor
 *  The overlap must be and integral divisor of the FFT Size.  The ambiguity
 *  shown by the error output is the modulus of the parameters.
 *
 *Window is too large!
 *  The window function size must be less-than-or-equal-to the FFT size.
 *
 *Window is smaller than FFT Size, assuming zero-padding.
 *  The program is assuming you intend to use a smaller window function for a
 *  large FFT.  This output is provided in case you accidentally provided the
 *  wrong window function.
 *
 *Error performing calculations.
 *  There was a gross error with the calculation routine.  This could be because
 *  the program ran out of memory, or there was some failure with the
 *  multithreading system
 *
 *ERROR; Cannot create message queue
 *  Is your system POSIX compliant?  This program requires the use of pthreads
 *  and POSIX message queues.  There was a problem creating the message queue.
 *  Investigate /dev/mqueue
 *
 *ERROR; return code from pthread_creat() is [xx]
 *  There was a serious problem creating child threads.  The program could have
 *  run out of system resources, or any number of reasons.
 *
 *Input data terminated with unaligned data
 *  It is entirely possible to have an input data file that is not some integral
 *  mutiple of (FFT Size)/Overlap.  If this is the case, the program runs until
 *  it doesn't have enough data to complete an FFT.  After encountering the tail
 *  of the file, the program finishes remaining calculations and exits
 *  gracefully.  This message is provided to indicate that data was lost in the
 *  FFT process (The bytes at the end of the file that didn't fit into the last
 *  FFT.
 *
 *
 *BENCHMARK MODE
 *
 *Complide with -DBENCHMARK to get Elapsed Time (ET) output of the realtime
 *used to calculate all the FFTs



Documentation for energycalculator:

 *This program takes I-Q data from a specified file, and computes a stream of energy bins.
 *
 *The I-Q Data is assumed to be stored in 'binary' complex<float> format.  (2 floats per 
 *sample). The energy bin size is defined as the number of samples summed to compute energy.
 *
 *This is the post-processing equivalent of usrp-energy.  This program can be run with data 
 *from usrp-record if usrp-record is configured to store fc32 data.
 *
 *The commandline options are:
 *
 * -o [file]       Output File  -The output file contains raw float data
 *                               representing the computed energy in each bin
 *
 * -b [size]       Bin Size     -Energy bin size in samples
 *
 * -i [file]	   Input File	-The input file contains raw float data


Documentation for fftcompute:
 *This program is designed to leverage single-precision fftw3 libraries and
 *pthreads to compute spectral periodigrams using an applied window function
 *and the magnitude of a 1D DFT.
 *
 *The commandline options are:
 *
 * -i [file]       Input File   -This input file is expected to contain
 *                               raw complex float data representing recorded
 *                               samples (from MATLAB, GNU Radio, etc.)
 *
 * -o [file]       Output File  -The output file contains raw float data
 *                               representing the computed spectral periodigram
 *                               of the recorded signal.  The peridigram is
 *                               computed with fftw3f's 1D DFT.
 *
 * -s [size]       FFT Size     -Size of the FFT to compute
 *
 * -l [overlap]    Overlap      -How many FFTs overlap in the space of a
 *                               transform.  That is, a transform of FFT Size is
 *                               computed every (FFT Size)/Overlap samples.
 *                               Sanity checking is performed to ensure that the
 *                               speficied overlap is appropriate
 *
 * -w [file]       Window File  -The window file is expected to contain raw
 *                               floats representing the window function to
 *                               apply to each FFT.  It is acceptable for the
 *                               window function to be smaller than the FFT.
 *                               If this is the case, then the program
 *                               left-justifies the window with respect to the
 *                               input array to fftw3f, and zero-pads the
 *                               remaining values.
 *
 * -c [children]   Child Threads-This specifies the number of child worker
 *                               threads to spawn for FFTCompute.  If your
 *                               processor has N cores (including hyperthreading
 *                               cores), it is recommended to specify N-1
 *                               children.  This allows for the parent thread
 *                               to process the input data for a total of N
 *                               threads.
 *
 *
 *
 *Description of error messages:
 *
 *Need at least one child thread
 *  Worker threads spawn from the parent process.  You must specify at least
 *  one child thread to do the FFT calculations.
 *
 *Cannot open window file
 *  There was a problem opening the window file.
 *
 *Incompatible FFT Size and Overlap factor
 *  The overlap must be and integral divisor of the FFT Size.  The ambiguity
 *  shown by the error output is the modulus of the parameters.
 *
 *Window is too large!
 *  The window function size must be less-than-or-equal-to the FFT size.
 *
 *Window is smaller than FFT Size, assuming zero-padding.
 *  The program is assuming you intend to use a smaller window function for a
 *  large FFT.  This output is provided in case you accidentally provided the
 *  wrong window function.
 *
 *Error performing calculations.
 *  There was a gross error with the calculation routine.  This could be because
 *  the program ran out of memory, or there was some failure with the
 *  multithreading system
 *
 *ERROR; Cannot create message queue
 *  Is your system POSIX compliant?  This program requires the use of pthreads
 *  and POSIX message queues.  There was a problem creating the message queue.
 *  Investigate /dev/mqueue
 *
 *ERROR; return code from pthread_creat() is [xx]
 *  There was a serious problem creating child threads.  The program could have
 *  run out of system resources, or any number of reasons.
 *
 *Input data terminated with unaligned data
 *  It is entirely possible to have an input data file that is not some integral
 *  mutiple of (FFT Size)/Overlap.  If this is the case, the program runs until
 *  it doesn't have enough data to complete an FFT.  After encountering the tail
 *  of the file, the program finishes remaining calculations and exits
 *  gracefully.  This message is provided to indicate that data was lost in the
 *  FFT process (The bytes at the end of the file that didn't fit into the last
 *  FFT.
 *
 *
 *BENCHMARK MODE
 *
 *Complide with -DBENCHMARK to get Elapsed Time (ET) output of the realtime
 *used to calculate all the FFTs
 *
 *On an Intel(R) Xeon(R) W3530 processor with 6GB of RAM, this program will
 *compute about 120,000 FFTs of size 1024 per second with 7 worker threads.
 *
 *This program is the post-processing equivalent of usrp-sensor.  usrp-record
 *data may be used if configured to record fc32 data




BUILDING:


To build this suite:

$mkdir build
$cd build
$cmake ..
$make

To build with debugging symbols and additional output:

$mkdir build
$cd build
$cmake -DDEBUG=1 ..
$make

You can change the wire format and host format for the various programs with:

To enable sc8 wire format (default sc16) (Affects usrp-energy usrp-recorder and
usrp-sensor)
-DWIRE_SC8=1

To enable sc16 host format (default fc32) (Only affects usrp-recorder)
-DHOST_SC16=1



ChangeLog
0.1 - Initial Release 2012
0.2 - Modification to usrp-sensor
0.3 - Change to cmake build system and release to github 20130701


include/fft_thread.h
/*Copyright 2013,2012 Joseph "Mitch" Davis mitchd@vt.edu
 *
 *This file is part of usrp-utils.
 *
 *   usrp-sensor is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   usrp-sensor is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with usrp-utils.  If not, see <http://www.gnu.org/licenses/>.
 *
 *
 *
 *This is the worker thread definition.  Used in the usrp-sensor program
 */
#ifndef FFT_THREAD_H_INCLUDED
#define FFT_THREAD_H_INCLUDED

#include <pthread.h>
#include <stdio.h>
#include <fftw3.h>
#include <math.h>
#include <complex.h>
#include <iostream>
#include <mqueue.h>


#define __FFT_THREAD_START       1
#define __FFT_THREAD_KILL        2
#define __FFT_THREAD_MSG_LENGTH  1
#define __FFT_THREAD_MSG_PRIO    0

struct fft_thread_data
{
    FILE*             outputFile;     //File to output the FFT results

    fftwf_plan        plan;           //fftw3 fft plan

    float complex*    outputData;     //fft output data (only valid after
    //fftw_execute)

    float complex*    inputData;      //fft input data (depending on the fft plan,
    //these data may get destroyed)

    float*            window;         //window function

    pthread_mutex_t*  output_mutex;   //Output file mutex

    int               fft_size;       //FFT Size
    volatile bool     is_running;     //Signal that the thread is busy
    char*             mq_name;        //mqueue name

    int               my_id;          //child thread id number
    volatile int*     next_thread;    //next thread id to execute
    int               max_children;   //maximum number of child processes
};

/*fft_thread_start
 *
 *pthread starting function for the fftw calculation thread
 *see the above struct for key parameters
 */
void* fft_thread_start( void* fft_thread_arg );


#endif // FFT_THREAD_H_INCLUDED


src/common/fft_thread.cpp
/*Copyright 2013,2012 Joseph "Mitch" Davis mitchd@vt.edu
 *
 *This file is part of usrp-utils.
 *
 *   usrp-utils is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   usrp-sensor is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with usrp-utils.  If not, see <http://www.gnu.org/licenses/>.
 *
 *
 *
 *
 *This is the worker thread implementation.  Used in the FFTCompute program
 */

#include "fft_thread.h"


void* fft_thread_start( void* fft_thread_arg )
{
    const int FLOAT_SIZE = sizeof(float);


    //Sort out the input data
    struct fft_thread_data* my_thread_data;

    my_thread_data = reinterpret_cast<fft_thread_data*>(fft_thread_arg);

    float *magnitude  = new float[my_thread_data->fft_size];

    //increment is used for adavancing the thread_control to the next thread
    //in line
    int   increment   = 1;
    if( my_thread_data->my_id == my_thread_data->max_children - 1 )
    {
        increment = -1*my_thread_data->my_id;
    }

    //Open up this child's message queue
    mqd_t my_queue = mq_open( my_thread_data->mq_name, O_RDONLY );

    //Variables used for interfacing with the mqueue
    char*     my_msg        = new char[255];
    const int my_msg_length = __FFT_THREAD_MSG_LENGTH;
    unsigned* my_prio       = NULL;


    while( true )
    {
        //mq_receive is blocking, so we'll sit here and idle until we get a
        //message
        mq_receive( my_queue, my_msg, my_msg_length, my_prio );

        //Check to see if we got the message to kill this thread
        if( my_msg[0] == static_cast<char>(__FFT_THREAD_KILL) )
            break;
        else if( my_msg[0] == static_cast<char>(__FFT_THREAD_START) )
        {
            //Got a message to signal an FFT computation

            //Signal that the thread is running
            my_thread_data->is_running = true;

            //Apply the window function
            for(int i = 0; i < my_thread_data->fft_size; i++ )
            {
                my_thread_data->inputData[i] *= my_thread_data->window[i];
            }

            //Compute fft
            fftwf_execute( my_thread_data->plan );

            //Compute magnitude (we don't want to store phase information)
            for(int i = 0; i < my_thread_data->fft_size; i++ )
            {
                magnitude[i] = cabsf( my_thread_data->outputData[i] );
            }

            //Ensure that we're writing data to the output file in order
            while( *(my_thread_data->next_thread) != my_thread_data->my_id )
                ;

            //Write results to the output file

            //Lock mutex in preparation to write
            //this is a normal mutex, so this will block until we can grab it
            pthread_mutex_lock( my_thread_data->output_mutex );

            //Write to file

            //Negative freqs first
            fwrite(magnitude+(my_thread_data->fft_size/2), FLOAT_SIZE,
                   my_thread_data->fft_size/2, my_thread_data->outputFile);
            //Positive freqs next
            fwrite(magnitude, FLOAT_SIZE,
                   my_thread_data->fft_size/2, my_thread_data->outputFile);

            //Advance the next_thread
            *(my_thread_data->next_thread) += increment;

            //Unlock mutex
            pthread_mutex_unlock( my_thread_data->output_mutex );

            //Signal that we're done running and going back to the idle state
            my_thread_data->is_running = false;
        }
    }
    //Kill signal received

    //Close the mqueue
    mq_close( my_queue );
    //Free memory
    delete [] my_msg;
    delete [] magnitude;
    //Kill thread
    pthread_exit(NULL);
}


src/energycalculator/energycalculator.cpp
/*Copyright 2013,2012 Joseph "Mitch" Davis mitchd@vt.edu
 *
 *This file is part of usrp-utils.
 *
 *   usrp-utils is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   usrp-energy is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with usrp-utils.  If not, see <http://www.gnu.org/licenses/>.
 *
 *
 *This program takes I-Q data from a specified file, and computes a stream of energy bins.
 *
 *The I-Q Data is assumed to be stored in 'binary' complex<float> format.  (2 floats per 
 *sample). The energy bin size is defined as the number of samples summed to compute energy.
 *
 *This is the post-processing equivalent of usrp-energy.  This program can be run with data 
 *from usrp-record if usrp-record is configured to store fc32 data.
 *
 *The commandline options are:
 *
 * -o [file]       Output File  -The output file contains raw float data
 *                               representing the computed energy in each bin
 *
 * -b [size]       Bin Size     -Energy bin size in samples
 *
 * -i [file]	   Input File	-The input file contains raw float data
 *
 * Changelog
 *
 * 0.1 - Initial release 2012
 * 0.3 - Change to cmake and upload to github 20130701
 *
 *
 */

#include <iostream>
#include <stdio.h>
#include <cstring>
#include <cstdlib>
#include <cstdio>
#include <cmath>
#include <unistd.h>
//Uncomment this to get gratuitous debug information
//#define DEBUG 1
using namespace std;

/*useage()
 *
 *Display program useage information
 *
 */
void useage();

/*calculateTask
 *
 *This is the main computation of the program.  inputFileName must be a valid
 *data file, and we must have write permission for outputFileName.
 *
 *energyBinSize defines the number of samples summed for computing the energy
 */
int calculateTask( char* inputFileName, char* outputFileName, int energyBinSize );


/*wrideData
 *
 *Writes 4 bytes to the specified ofstream object
 *
 */
int writeData( FILE* outputFile, double data );



int main( int argc, char* argv[])
{
  //Ensure the correct number of arguments were passed
  if( argc != 7){
    cout << "Only " << argc << " parameters entered" << endl;
    useage();
    return -1;
  }

  char* inputFileName = NULL;
  char* outputFileName = NULL;
  int   energyBinSize = 0;
  int   arg = 0;

  //argument parsing
  while( (arg = getopt( argc, argv, "i:o:s:")) != -1 ){
#ifdef DEBUG
    cout << "Arg: " << optarg << endl;
#endif
    switch (arg){

    case 'i':
      inputFileName = new char[sizeof(optarg)];
      strcpy(inputFileName,optarg);
#ifdef DEBUG
      cout << "Input Filename: " << inputFileName << endl;
#endif
      break;

    case 'o':
      outputFileName = new char[sizeof(optarg)];
      strcpy(outputFileName,optarg);
#ifdef DEBUG
      cout << "Output Filename: " << outputFileName << endl;
#endif
      break;

    case 's':
      energyBinSize = atoi(optarg);
#ifdef DEBUG
      cout << "Energy Bin Size: " << energyBinSize << endl;
#endif
      break;

    case '?':
      useage();
      if( inputFileName )
        delete [] inputFileName;
      if( outputFileName )
        delete [] outputFileName;
      return -1;
    }
  }

  if( !calculateTask( inputFileName, outputFileName, energyBinSize ) ){
    cout << "Error performing calculations" << endl;
    delete [] inputFileName;
    delete [] outputFileName;
    return 1;
  }

  delete [] inputFileName;
  delete [] outputFileName;
  return 0;
}


void useage( )
{
  cout  << "Useage:\t " << endl
        << "-i <file>\t Input File" << endl
        << "-o <file>\t Output File" << endl
        << "-s <size>\t Energy Bin Size" << endl;
}


int calculateTask( char* inputFileName, char* outputFileName, int energyBinSize )
{
  const int FLOAT_SIZE = 4; //Number of bytes in a single-precision float

  FILE* inputFile;
  FILE* outputFile;

  inputFile = fopen( inputFileName, "r");
#ifdef DEBUG
  cout << "Input file open attempt." << endl;
#endif
  outputFile = fopen( outputFileName, "w" );
#ifdef DEBUG
  cout << "Output file open attempt." << endl;
#endif

  /*Make sure the files actually opened.
   *
   *If not... gracefully exit.
   *We should really be throwing exceptions... but I'm feeling lazy
   */
  if( !inputFile || !outputFile){
#ifdef DEBUG
    cout << "Input File " << inputFileName << " Open? " << (inputFile ? "True" : "False") << endl;
    cout << "Output File " << outputFileName << " Open? " << (outputFile ? "True" : "False") << endl;
#endif
    if( inputFile )
      fclose( inputFile );
    if( outputFile )
      fclose( outputFile );
    return 0;
  }
#ifdef DEBUG
  cout << "Input/Output files opened successfully." << endl;
#endif

  float i = 0;
  float q = 0;

  double mag = 0;
  int    iterator = 0;

  while( !feof(inputFile) ){
      //Read in the I-Q of one sample... FLOAT_SIZE bytes per float
      fread( &i, FLOAT_SIZE, 1, inputFile );
      fread( &q, FLOAT_SIZE, 1, inputFile );

      //Ensure we're calling the double version of pow, otherwise data is lost
      mag += pow(static_cast<double>(i),2.) + pow(static_cast<double>(q),2.);

      //Reached the end of the bin?
      if( ++iterator == energyBinSize){
        iterator = 0;
        //If so, write 4 bytes
        if( !writeData( outputFile, mag ) ){
          fclose(inputFile);
          fclose(outputFile);
          return 0;
        }
        mag = 0;
      }
  }
  //Toss out any leftovers (incomplete energy bin)
  fclose(inputFile);
  fclose(outputFile);
#ifdef DEBUG
  cout << endl << "Files Closed." << endl;
#endif
  return 1;
}

int writeData( FILE* outputFile, double data )
{
  const int DOUBLE_SIZE = 8; //Number of bytes in a single-precision float
  fwrite(&data, DOUBLE_SIZE, 1, outputFile);
#ifdef DEBUG
  cout << ".";
#endif
  return 1;
}
